tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -38,131 +40,161 @@ import           Source.Span
+    let spanStart = pointToPos (nodeStartPoint node)
+        spanEnd   = pointToPos (nodeEndPoint node)
+    pure (Span spanStart spanEnd)
 
 pointToPos :: TSPoint -> Pos
 pointToPos (TSPoint line column) = Pos (fromIntegral line) (fromIntegral column)
 
-instance Unmarshal a => Unmarshal (Maybe a) where
-  unmarshalNodes [] = pure Nothing
-  unmarshalNodes listOfNodes = Just <$> unmarshalNodes listOfNodes
-
-instance (Unmarshal (f a), Unmarshal (g a), SymbolMatching (f a), SymbolMatching (g a)) => Unmarshal ((f :+: g) a) where
-  unmarshalNodes [node] = do
-    let lhsSymbolMatch = symbolMatch (Proxy @(f a)) node
-        rhsSymbolMatch = symbolMatch (Proxy @(g a)) node
-    if lhsSymbolMatch
-      then L1 <$> unmarshalNodes @(f a) [node]
-      else if rhsSymbolMatch
-        then R1 <$> unmarshalNodes @(g a) [node]
-        else fail $ showFailure (Proxy @((f :+: g) a)) node
-  unmarshalNodes [] = fail "expected a node of type ((f :+: g) a) but didn't get one"
-  unmarshalNodes _ = fail "expected a node of type ((f :+: g) a) but got multiple"
-
-
-instance Unmarshal a => Unmarshal [a] where
-  unmarshalNodes (x:xs) = do
-    head' <- unmarshalNodes [x]
-    tail' <- unmarshalNodes xs
-    pure $ head' : tail'
-  unmarshalNodes [] = pure []
 
+-- | Optional/repeated fields occurring in product datatypes are wrapped in type constructors, e.g. 'Maybe', '[]', or 'NonEmpty', and thus can unmarshal zero or more nodes for the same field name.
+class UnmarshalField t where
+  unmarshalField
+    :: ( Carrier sig m
+       , Member (Reader ByteString) sig
+       , Member (Reader (Ptr Cursor)) sig
+       , MonadFail m
+       , MonadIO m
+       , Unmarshal f
+       , UnmarshalAnn a
+       )
+    => [Node]
+    -> m (t (f a))
+
+instance UnmarshalField Maybe where
+  unmarshalField []  = pure Nothing
+  unmarshalField [x] = Just <$> unmarshalNode x
+  unmarshalField _   = fail "expected a node of type ((f :+: g) a) but got multiple"
+
+instance UnmarshalField [] where
+  unmarshalField (x:xs) = do
+    head' <- unmarshalNode x
+    tail' <- unmarshalField xs
+    pure $ head' : tail'
+  unmarshalField [] = pure []
 
-instance Unmarshal a => Unmarshal (NonEmpty a) where
-  unmarshalNodes (x:xs) = do
-    head' <- unmarshalNodes [x]
-    tail' <- unmarshalNodes xs
+instance UnmarshalField NonEmpty where
+  unmarshalField (x:xs) = do
+    head' <- unmarshalNode x
+    tail' <- unmarshalField xs
     pure $ head' :| tail'
-  unmarshalNodes [] = fail "expected a node but didn't get one"
+  unmarshalField [] = fail "expected a node but didn't get one"
+
 
-class SymbolMatching a where
+class SymbolMatching (a :: * -> *) where
   symbolMatch :: Proxy a -> Node -> Bool
 
   -- | Provide error message describing the node symbol vs. the symbols this can match
   showFailure :: Proxy a -> Node -> String
 
-instance SymbolMatching a => SymbolMatching (Maybe a) where
-  symbolMatch _ = symbolMatch (Proxy @a)
-  showFailure _ = showFailure (Proxy @a)
```

Comment:
A bunch of these `SymbolMatching` instances are completely unnecessary now: we only use `SymbolMatching` for sums, and these modifier types only occur in products.